website/public_adventures.py

```diff
@@ -119,35 +62,85 @@ def filtering(self, user, index_page=False):
                 tags = [_tag for _tag in toReset.split(",") if _tag != tag]
             else:
                 tags = tag.split(",")
+            tags = [t for t in tags if t]
+
+        # Get indexes
+        level_lang_adventure_ids, result = self.db.get_paginated_indexes(
+            {"field_value": f"level_lang_{level}_{language}"},
+            pagination_token=page, limit=20)
+        # In case we need want any language, we retrieve by the main level index.
+        if not level_lang_adventure_ids:
+            level_lang_adventure_ids, result = self.db.get_paginated_indexes({"field_value": f"level_{level}"},
+                                                                             pagination_token=page, limit=20)
+        prev_page_token = result.prev_page_token
+        next_page_token = result.next_page_token
+
+        tag_adventure_ids = set()
+        for _t in tags:
+            tag_adventure_ids.update(self.db.get_public_adventures_indexes({"field_value": f"tag_{_t}"}))
 
-            tags = [_tag for _tag in tags if _tag]  # filter out empty strings.
-            for _tag in tags:
-                # In case a selected set of adventures doesn't have the given tag to filter on,
-                # we decide that that tag cannot be used for filtering.
-                if not any(_tag in adv.get("tags") for adv in adventures):
-                    tags.remove(_tag)
-            if tags:
-                adventures = [adv for i, adv in enumerate(adventures) if any(tag in tags for tag in adv.get("tags"))]
-            # adjust available languages after fitlering on tags.
-            self.update_filters(adventures, "lang")
-
-        if search:
-            adventures = [adv for adv in adventures if search.lower() in adv.get("name").lower()]
-            self.update_filters(adventures, "lang")
-            self.update_filters(adventures, "tag")
+        tag_adventure_ids.intersection_update(level_lang_adventure_ids)
+        adventure_ids = level_lang_adventure_ids.union(level_lang_adventure_ids, tag_adventure_ids)
+
+        adventures = self.db.batch_get_adventures(adventure_ids)
 
         initial_tab = None
         initial_adventure = None
         commands = {}
         prev_level = None
         next_level = None
+
+        customized_adventures = []
+        included = {}
+        adventures = list(adventures.values())
+        for adventure in adventures:
+            if not adventure:
+                continue
+            if language and adventure.get("language", g.lang) != language:
+                continue
+            if tags and not any(_t in adventure.get("tags", []) for _t in tags):
+                continue
+            if search and not search.lower() in adventure.get("name").lower():
+                continue
```

Comment:
Filtering should already have happened earlier, otherwise the pagination is not going to do what you want.

Your page may end up with `0` items and a `Next >>` button, which looks very weird.